### PR TITLE
CTM-394: Bypass passport validation for public NRES snapshots

### PR DESCRIPTION
**Pull request by snf2ye**
Ticket: https://broadworkbench.atlassian.net/browse/CTM-394

## Summary
Adds support for bypassing RAS passport validation when a snapshot has NRES consent code and is publicly accessible via SAM reader policy.

This implements a special case where:
1. Snapshot has `NRES` consent code (case-insensitive)
2. Snapshot's SAM reader policy is public

When both conditions are met, passport validation is bypassed and the event is logged at INFO level.

## Changes
- **SnapshotSummary.java**: Add `NRES_CONSENT_CODE` constant and `isPublicConsentCode()` helper method
- **SnapshotService.java**: Add `canBypassPassportValidation()` method and update `verifyPassportAuth()` to support bypass with bearer token
- **Tests**: Comprehensive unit tests for bypass logic, edge cases, and integration scenarios

## Security Considerations
- Requires BOTH NRES consent code AND public reader policy
- Falls back to normal validation if bearer token unavailable
- Logs all bypass events for audit purposes
- Case-insensitive NRES matching for robustness

## Test Plan
- [x] Unit tests for `isPublicConsentCode()` (case-insensitive)
- [x] Unit tests for `canBypassPassportValidation()` 
- [x] Unit tests for bypass in `verifyPassportAuth()`
- [x] Tests for private NRES requiring validation
- [x] Tests for non-NRES always requiring validation
- [x] DRS integration tests with public NRES
- [x] All existing tests pass

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/main/java/bio/terra/service/auth/iam/IamService.java [hunk 1]
```diff
@@ -633,6 +633,15 @@ public boolean getPolicyPublicV2(
...
 
+  public boolean getPolicyPublicV2AsSA(
+      IamResourceType iamResourceType, UUID resourceId, String policyName) {
+    String tdrSaAccessToken = googleCredentialsService.getApplicationDefaultAccessToken(SCOPES);
+    return callProvider(
+        () ->
+            iamProvider.getPolicyPublicV2(
+                tdrSaAccessToken, iamResourceType, resourceId, policyName));
+  }
+
```
Comment (snf2ye) [inline src/main/java/bio/terra/service/auth/iam/IamService.java:641]: In my local testing, this call is failing with the following error. My best guess is that it has something to do with my local setup - Perhaps I'm not authenticating as the TDR SA when I run the service? We make similar calls with the TDR SA Access token across this service, so I don't really think it is actually broken. 

Proposed path forward: The way this is failing means that everything should continue to work as it does today: Even public snapshots will require a valid passport that grants them access. So, it should not pose a regression. I should just be able to iterate on this change without needing to revert if it does not work on dev. 

> 2026-04-23 15:31:51,750 WARN  [drs-resolution-thread-1] b.t.service.snapshot.SnapshotService: Error checking public status for snapshot 48d3fc6d-f066-441c-b4e9-8fae0bf759a6, cannot bypass passport validation
> bio.terra.service.auth.iam.exception.IamNotFoundException@89c993a[causes=[],statusCode=404 NOT_FOUND]
> 	at bio.terra.service.auth.iam.sam.SamIam.convertSamExToDataRepoEx(SamIam.java:1068)
> 	at bio.terra.service.auth.iam.sam.SamRetry.handleApiException(SamRetry.java:93)
> 	at bio.terra.service.auth.iam.sam.SamRetry.perform(SamRetry.java:62)
> 	at bio.terra.service.auth.iam.sam.SamRetry.retry(SamRetry.java:46)
> 	at bio.terra.service.auth.iam.sam.SamIam.getPolicyPublicV2(SamIam.java:985)
> 	at bio.terra.service.auth.iam.IamService.lambda$getPolicyPublicV2AsSA$40(IamService.java:641)
> 	at bio.terra.service.auth.iam.IamService.callProvider(IamService.java:90)
> 	at bio.terra.service.auth.iam.IamService.getPolicyPublicV2AsSA(IamService.java:639)
> 	at bio.terra.service.snapshot.SnapshotService.isSnapshotPublic(SnapshotService.java:885)
> 	at bio.terra.service.snapshot.SnapshotService.canBypassPassportValidation(SnapshotService.java:875)
> 	at bio.terra.service.snapshot.SnapshotService.verifyPassportAuth(SnapshotService.java:905)
> 	at bio.terra.service.filedata.DrsService.verifyPassportAuth(DrsService.java:422)
> 	at bio.terra.service.filedata.DrsService.lambda$lookupObjectByDrsIdPassport$0(DrsService.java:249)
> 	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
> 	at java.base/java.lang.Thread.run(Unknown Source)
> Caused by: org.broadinstitute.dsde.workbench.client.sam.ApiException: Message:
> HTTP response code: 404
> HTTP response body: {"causes":[],"message":"Resource datasnapshot/48d3fc6d-f066-441c-b4e9-8fae0bf759a6 not found","source":"sam","stackTrace":[],"statusCode":404}

File: tools/setupResourceScripts/files/V2F_GWAS_Summary_Statistics/all_data_types.json [hunk 2]
```diff
@@ -3,3 +3,4 @@
...
 {"variant_id":"4:97820191:G:C","bool_column":false, "date_column":"2022-7-04","date_array_column":["2022-7-02","2022-7-03","2022-7-04"],"date_time_column":"2022-07-05 12:00:01","date_time_array_column":["2022-07-05 12:00:01"],"float_column":123.002930495093854,"float_array_column":[1.232,123.00],"float64_column":123.002930495093854,"float64_array_column":[1.232,123.00],"int_column":2314,"int_array_column":[245,3325,2343],"int64_column":234235,"int64_array_column":[1,234235],"numeric_column":234,"numeric_array_column":[2,2,3],"time_column":"23:59:59.999999","time_array_column":["1:34:01","12:34:01.0908"],"timestamp_column":"2023-06-01 01:02:40","timestamp_array_column":["2023-06-02 01:04:40"]}
+{"variant_id": "123", "file_ref_column": {"sourcePath": "gs://broad-jade-dev-data-bucket/rori.json", "targetPath":  "/test/rori.json", "mimeType": "application/json"}}
```
Comment (snf2ye) [inline tools/setupResourceScripts/files/V2F_GWAS_Summary_Statistics/all_data_types.json]: It has been a real painpoint that we don't have a test file ingested in the setup script. This adds one!

File: src/main/java/bio/terra/service/auth/iam/IamService.java [hunk 3]
```diff
@@ -633,6 +633,15 @@ public boolean getPolicyPublicV2(
         () -> iamProvider.getPolicyPublicV2(token, iamResourceType, resourceId, policyName));
   }
 
+  public boolean getPolicyPublicV2AsSA(
+      IamResourceType iamResourceType, UUID resourceId, String policyName) {
+    String tdrSaAccessToken = googleCredentialsService.getApplicationDefaultAccessToken(SCOPES);
+    return callProvider(
+        () ->
+            iamProvider.getPolicyPublicV2(
+                tdrSaAccessToken, iamResourceType, resourceId, policyName));
```
Comment (davidangb) [inline src/main/java/bio/terra/service/auth/iam/IamService.java]: Your call, this is mostly style: this could delegate to `getPolicyPublicV2()` just above it instead of re-implementing the call:
```java
    String tdrSaAccessToken = googleCredentialsService.getApplicationDefaultAccessToken(SCOPES);
    return getPolicyPublicV2(tdrSaAccessToken, iamResourceType, resourceId, policyName);
```

File: src/main/java/bio/terra/service/snapshot/SnapshotSummary.java [hunk 4]
```diff
@@ -226,4 +228,15 @@ private List<StorageResourceModel> toStorageResourceModel() {
...
+  /**
+   * @param model a SnapshotSummaryModel
+   * @return whether the snapshot has a public consent code (NRES) that allows bypass of passport
+   *     validation when the snapshot is also publicly accessible
+   */
+  public static boolean isPublicConsentCode(SnapshotSummaryModel model) {
+    return !StringUtils.isBlank(model.getPhsId())
+        && !StringUtils.isBlank(model.getConsentCode())
+        && NRES_CONSENT_CODE.equalsIgnoreCase(model.getConsentCode());
+  }
```
Comment (davidangb) [inline src/main/java/bio/terra/service/snapshot/SnapshotSummary.java:241]: I'm not sure what is stored in the consent code field in practice ... is it possible to have a consent code like `HMB NRES GRU`? I realize that consent code doesn't actually make sense, but still ...

File: tools/setupResourceScripts/files/V2F_GWAS_Summary_Statistics/dataset_schema.json [hunk 5]
```diff
@@ -904,7 +904,7 @@
           },
           {
             "name": "file_ref_column",
-            "datatype": "dirref"
+            "datatype": "fileref"
```
Comment (davidangb) [inline tools/setupResourceScripts/files/V2F_GWAS_Summary_Statistics/dataset_schema.json:907]: this seems correct, but is it intentionally part of this PR?

**Review by davidangb — COMMENTED** [review]
minor comments/questions:

File: src/main/java/bio/terra/service/snapshot/SnapshotSummary.java (hunk 4 above)
Comment (snf2ye) [reply src/main/java/bio/terra/service/snapshot/SnapshotSummary.java:241]: That's a great question. 
I just took a look at the consent codes that exist today on snapshots and it looks like NRES only appears on it's own. My only concern with checking if the consent code contains "NRES" is in the case where they do something silly and have something like "XNRES" and we inadvertently make something public access. Given that NRES on its own should work for existing use cases and is the more strict interpretation, I'm included to leave it as it is and we could always open it up in the future if there are combination consent codes.

File: tools/setupResourceScripts/files/V2F_GWAS_Summary_Statistics/dataset_schema.json (hunk 5 above)
Comment (snf2ye) [reply tools/setupResourceScripts/files/V2F_GWAS_Summary_Statistics/dataset_schema.json:907]: Yes, it is drive-by change, but together with the change to all_data_types.json, it enables a file to be uploaded in the setup script so that we can test DRS URIs.

File: src/main/java/bio/terra/service/auth/iam/IamService.java (hunk 1 above)
Comment (calypsomatic) [reply src/main/java/bio/terra/service/auth/iam/IamService.java:641]: I'm good with calling this local setup and merging.  Or is there any chance you'd get a different result if you changed it up like David suggested in his comment?

File: src/main/java/bio/terra/service/auth/iam/IamService.java (hunk 3 above)
Comment (snf2ye) [reply src/main/java/bio/terra/service/auth/iam/IamService.java]: Oh duh, that makes a lot of sense. Changed!

File: src/main/java/bio/terra/service/auth/iam/IamService.java (hunk 1 above)
Comment (snf2ye) [reply src/main/java/bio/terra/service/auth/iam/IamService.java:641]: Are you referring to this comment? https://github.com/DataBiosphere/jade-data-repo/pull/2094#discussion_r3132104943
If so, I don't think that should impact the result here. It _should_ be a swap of effectively the same code.

**Review by davidangb — APPROVED** [review]

**Review by calypsomatic — APPROVED** [review]
